src/VisualStudio/Core/Def/Implementation/LanguageService/AbstractPackage.cs

```diff
@@ -1,23 +1,24 @@
-﻿using System;
-using System.Collections.Generic;
-using System.Linq;
-using System.Text;
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
 using System.Threading;
-using System.Threading.Tasks;
 using System.Windows.Threading;
 using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
 using Microsoft.CodeAnalysis.Utilities;
 using Microsoft.VisualStudio.Shell;
+using Task = System.Threading.Tasks.Task;
 
 namespace Microsoft.VisualStudio.LanguageServices.Implementation.LanguageService
 {
-    internal abstract class AbstractPackage : Package
+    internal abstract class AbstractPackage : AsyncPackage
     {
         protected ForegroundThreadAffinitizedObject ForegroundObject;
 
-        protected override void Initialize()
+        protected override async Task InitializeAsync(CancellationToken cancellationToken, IProgress<ServiceProgressData> progress)
         {
-            base.Initialize();
+            await base.InitializeAsync(cancellationToken, progress).ConfigureAwait(true);
+
+            await JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
 
             // Assume that we are being initialized on the UI thread at this point, and setup our foreground state
             var kind = ForegroundThreadDataInfo.CreateDefault(ForegroundThreadDataKind.ForcedByPackageInitialize);
```

Comment:
Are we concerned that this is now running somewhat asynchronously, and might be backed up by other things loaded via MEF that expected this to be initialized?

Reply:
It's true that MEF parts can be activated independently of any VS Package's initialization. That being the case, no assumptions should be made about the package having been loaded from something that isn't itself directly a spawn from that Package. 
If this breaks something, that thing was asking to be broken. Not to say "go ahead and break it" but if you can call out areas that should be checked to reduce risk that this PR breaks something, please do.

Reply:
I think this was the only particular red flag I saw. Generally, the MEF vs. Package/services world has always been a thorn in Roslyn's side and I wouldn't be surprised in the least if a few assumptions snuck in that we haven't yet discovered over all the years.

Reply:
what happen if other MEF import VSWorkspace before ForegroundThreadDataInfo is initialized? are we guarded from that case?

Reply:
now, if AsyncPackage and adaption of JoinableFactory, we might not need these "LoadComponentsInUIContextOnceSolutionFullyLoaded" anymore. since now with JTF, we can just lazily initialize those when first used naturally.

Reply:
> what happen if other MEF import VSWorkspace before ForegroundThreadDataInfo is initialized? are we guarded from that case?

Is this a general question or something specific to this pull request? I wasn't intending to make changes that would impact this scenario.

> now, if AsyncPackage and adaption of JoinableFactory, we might not need these "LoadComponentsInUIContextOnceSolutionFullyLoaded" anymore. since now with JTF, we can just lazily initialize those when first used naturally.

We aren't fully moving to JTF at this point, but hopefully these are things we get the chance to clean up in the future.

Reply:
it was more general question. I would check in and see what breaks rather than guessing. but I do wonder whether what happen if someone (like winform) get VSWorkspace and start using it while Roslyn packages are not fully initialized yet. or OB is used while we are not fully initialized yet. or AsyncPackage makes sure that can't happen..

or it is possible but so rare, it doesn't matter.